credential.c

```diff
@@ -323,6 +332,20 @@ static int run_credential_helper(struct credential *c,
 
 	if (want_output) {
 		int r;
+		if (select_timeout.tv_sec) {
+			fd_set readfds;
+			FD_ZERO(&readfds);
+			FD_SET(helper.out, &readfds);
+
+			select(1, &readfds, NULL, NULL, &select_timeout);
+
+			if (!FD_ISSET(helper.out, &readfds)) {
+				/* Timeout complete before helper.out has bytes to read. */
+				kill_child_command(&helper);
```

Comment:
This bothers me a bit.  I'm wondering if we should just insert a `kill()` here and then let the existing call to `finish_command()` at the bottom wait for the child.  The existing finish code will end up calling `wait_or_whine()` which normalizes the child's exit code, so we'll be consistent.  And then the `child_exit`
and other cleanup.

You could also add a trace2_data after the `kill()` to log that we sent the signal, but that is not strictly necessary.

Reply:
Also, then you won't need the `kill_child_command()` function at all.  That might make upstream integration easier.

Reply:
Interesting. I hadn't considered that, because I expected the behavior to be "kill the child then leave the method" and not to go to a later `finish_command()`. I will play with this when I have a better testing environment ready.